### PR TITLE
Initial OIDC support (Google/GitHub/CircleCI -> Amazon Bedrock & Azure OpenAI)

### LINKED ISSUE #3505: [Feature]: Upstream OIDC for Azure OpenAI
### The Feature

Building upon #3499, I'll add support for Azure OpenAI.

### Motivation, pitch

This is the role definitions with the least amount of priv needed:

```json

{
    "properties": {
        "roleName": "invoke-openai",
        "description": "",
        "assignableScopes": [
            "/subscriptions/SUB_ID_HERE/resourceGroups/openai-demos"
        ],
        "permissions": [
            {
                "actions": [],
                "notActions": [],
                "dataActions": [
                    "Microsoft.CognitiveServices/accounts/OpenAI/deployments/audio/action",
                    "Microsoft.CognitiveServices/accounts/OpenAI/deployments/search/action",
                    "Microsoft.CognitiveServices/accounts/OpenAI/deployments/completions/action",
                    "Microsoft.CognitiveServices/accounts/OpenAI/deployments/chat/completions/action",
                    "Microsoft.CognitiveServices/accounts/OpenAI/deployments/extensions/chat/completions/action",
                    "Microsoft.CognitiveServices/accounts/OpenAI/deployments/embeddings/action",
                    "Microsoft.CognitiveServices/accounts/OpenAI/images/generations/action"
                ],
                "notDataActions": []
            }
        ]
    }
}
```

### Twitter / LinkedIn details

https://www.linkedin.com/in/davidmanouchehri/

### PR DESCRIPTION
**Pull request by Manouchehri**
## Title

Implement OIDC for Azure OpenAI and Amazon Bedrock requests.

Currently, this supports using OIDC from:

- Google Cloud Run (tested)
- GitHub Actions (untested but pretty sure it works)
- Circle CI (untested but should work fine)

And this supports using the auth token for:

- Amazon Bedrock
- Azure OpenAI

## Relevant issues

Resolves #3505 and #3401.

## Type

🆕 New Feature
🐛 Bug Fix
💻 Development Environment
🚄 Infrastructure

## Changes

This allows using OIDC with Azure OpenAI.

## Testing

```yaml
model_list:
  - model_name: gpt-4-0125-preview
    litellm_params:
      model: azure/gpt-4-0125-preview
      azure_ad_token: "oidc/google/https://example.com"
      api_version: "2024-03-01-preview"
      api_base: "https://demo-here.openai.azure.com"
    model_info:
      base_model: azure/gpt-4-0125-preview

  - model_name: claude-3-haiku-20240307
    litellm_params:
      model: bedrock/anthropic.claude-3-haiku-20240307-v1:0
      aws_region_name: us-east-1
      aws_session_name: "litellm"
      aws_role_name: "arn:aws:iam::YOUR_THING_HERE:role/litellm-google-demo"
      aws_web_identity_token: "oidc/google/https://example.com"

```

You must set the following two env vars:

```
AZURE_CLIENT_ID=asdf
AZURE_TENANT_ID=asdf
```

And see https://github.com/BerriAI/litellm/pull/3499 for the AWS IAM info.

## Notes

Docs will be in a later PR.

## Pre-Submission Checklist (optional but appreciated):

- [ ] I have included relevant documentation updates (stored in /docs/my-website)

## OS Tests (optional but appreciated):

- [ ] Tested on Windows
- [ ] Tested on MacOS
- [X] Tested on Linux

### CODE REVIEW

File: litellm/router.py [hunk 1]
```diff
@@ -2089,6 +2090,10 @@ def set_client(self, model: dict):
                     raise ValueError(
                         f"api_base is required for Azure OpenAI. Set it on your config. Model - {model}"
                     )
+                azure_ad_token = litellm_params.get("azure_ad_token")
+                if azure_ad_token is not None:
+                    if azure_ad_token.startswith("oidc/"):
```
Comment (ghost) [inline litellm/router.py:2095]: @Manouchehri can we add a simple unit test in `test_router.py` to make sure this is always respected? 

We've had previous regressions due to untested flows (thinking of cloudflare). Would help to have something for this

File: litellm/tests/test_secret_manager.py [hunk 2]
```diff
@@ -23,3 +23,36 @@ def test_aws_secret_manager():
...
+
+
+@pytest.mark.skipif(os.environ.get('CIRCLE_OIDC_TOKEN') is None, reason="Cannot run without being in a CircleCI Runner")
+def test_oidc_circleci():
+    secret_val = get_secret("oidc/circleci/https://bedrock-runtime.us-east-1.amazonaws.com/model/amazon.titan-text-express-v1/invoke")
+
+    print(f"secret_val: {redact_oidc_signature(secret_val)}")
+
+
+@pytest.mark.skipif(os.environ.get('CIRCLE_OIDC_TOKEN_V2') is None, reason="Cannot run without being in a CircleCI Runner")
```
Comment (ghost) [inline litellm/tests/test_secret_manager.py:54]: is there a way for us to add unit tests that *can* run on our ci/cd? @Manouchehri

Comment (Manouchehri) [reply litellm/tests/test_secret_manager.py:54]: The Circle CI ones should already work automatically!

Comment (ghost) [reply litellm/tests/test_secret_manager.py:54]: oh - how? we don't have these tokens

Comment (Manouchehri) [reply litellm/tests/test_secret_manager.py:54]: You do! :) 

https://app.circleci.com/pipelines/github/BerriAI/litellm/9535/workflows/f221c9b2-008d-47c0-b7cd-507d7e857d6c/jobs/20702?invite=true#step-99-2233_32

Comment (Manouchehri) [reply litellm/tests/test_secret_manager.py:54]: They are set by default. That's part of the ODIC magic.

Comment (ghost) [reply litellm/tests/test_secret_manager.py:54]: oh sweet!

**Pull request merged**